### DISCUSSION TITLE
[Feature Request] support receiving `respond` events when using an interpreter to run a machine

### DISCUSSION THREAD

**Discussion by AlCalzone** [statelyai/xstate#1470] [General] [answered]
When executing a machine using the built-in interpreter, I cannot find a way to receive events that were created by the machine using the `respond` action, e.g.:
```
App (using interpreter) ----EVENT---→   Machine
                                           ↓
                                     react to event
                                           ↓
App (   interpreter   ) ←--RESPONSE-- respond(...)
```
IMO it would make sense to treat the interpreter as the origin of the original event and this would give some flexibility to run a machine either standalone or spawned from another machine.

**Comment by davidkpiano** [comment, marked as answer]
Can you prepare a code example with a comment of what you're asking? Code examples are really helpful when thinking through problems like this.

[CodeSandbox (XState template)](https://codesandbox.io/s/xstate-example-template-qcwed)

**Reply by AlCalzone** [reply]
Ok, here you go:
https://codesandbox.io/s/example-for-receiving-respond-events-qt775

**Reply by Andarist** [reply]
Those events are given to the originating machine in this situation and you can act upon them:
https://codesandbox.io/s/example-for-receiving-respond-events-forked-guxzy?file=/src/index.js

Not sure though if this is what you are looking for. If not - could you describe your issue in more detail?

**Reply by AlCalzone** [reply]
The use case is basically that I want to pass a message to a machine. If that machine doesn't handle the message, it must be handled somewhere else and nothing should happen inside that machine.

I just tried your example without the transition to error and I also get the event back. I could have sworn that didn't work the last time I tried it...
However, the listener is now called for every transition and event inside the machine. I think it would be more idiomatic (and maybe more performant) if there was a listener that was only called for responses (`respond` and/or `sendParent` actions).

**Reply by Andarist** [reply]
Well, `respond` is actually just a *very thin* utility, meant to be used mainly for convenience purposes. It's not any separate concept - just look at its implementation:
https://github.com/davidkpiano/xstate/blob/4c3a274a8c9777e8076c4538fedbfc90966b5d87/packages/core/src/actions.ts#L298-L312
It's just a `send` in disguise 😉 So it's very unlikely that we'd like to special case this with an extra listener set or anything like that.

**Reply by AlCalzone** [reply]
Got it, thanks for your responses!
